ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -457,17 +429,94 @@ public Builder valueColumns(final Iterable<? extends Column> column) {
     }
 
     public LogicalSchema build() {
-      final List<Column> suppliedKey = keyBuilder.build();
+      final ImmutableList.Builder<NamespacedColumn> allColumns = ImmutableList.builder();
+
+      if (addImplicitRowTime) {
+        allColumns.add(IMPLICIT_TIME_COLUMN);
+      }
+
+      if (addImplicitRowKey) {
+        allColumns.add(IMPLICIT_KEY_COLUMN);
+      }
+
+      allColumns.addAll(explicitColumns.build());
+
+      return new LogicalSchema(allColumns.build());
+    }
+  }
+
+  private enum Namespace {
+    META,
+    KEY,
+    VALUE
+  }
+
+  @Immutable
+  private static final class NamespacedColumn {
+
+    private final Column column;
+    private final Namespace namespace;
+
+    static NamespacedColumn of(
+        final Column column,
+        final Namespace namespace
+    ) {
+      return new NamespacedColumn(column, namespace);
+    }
+
+    private NamespacedColumn(
+        final Column column,
+        final Namespace namespace
+    ) {
+      this.column = requireNonNull(column, "column");
+      this.namespace = requireNonNull(namespace, "namespace");
+    }
+
+    Column column() {
+      return column;
+    }
+
+    Namespace namespace() {
+      return namespace;
+    }
+
+    NamespacedColumn withSource(final SourceName sourceName) {
+      return NamespacedColumn.of(column.withSource(sourceName), namespace);
+    }
+
+    NamespacedColumn noSource() {
+      return NamespacedColumn.of(Column.of(column.name(), column.type()), namespace);
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      final NamespacedColumn that = (NamespacedColumn) o;
+      return Objects.equals(column, that.column)
+          && namespace == that.namespace;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(column, namespace);
+    }
+
+    @Override
+    public String toString() {
+      return toString(FormatOptions.none());
+    }
 
-      final List<Column> key = suppliedKey.isEmpty()
-          ? IMPLICIT_KEY_SCHEMA
-          : suppliedKey;
+    public String toString(final FormatOptions formatOptions) {
+      final String postFix = namespace == Namespace.VALUE
+          ? ""
+          : " " + namespace;
```

Comment:
did we intend `ROWTIME META` to be the toString value?

Reply:
It's not at the moment, as meta fields are excluded from `toString`.